src/main/java/org/reaktivity/nukleus/http2/internal/stream/Http2ServerFactory.java

```diff
@@ -1146,40 +1162,73 @@ private void onNetworkEnd(
                 }
                 else
                 {
-                    //streams.values().forEach(s -> s.doRequestAbortIfNecessary(traceId, authorization));
-                    streams.values().forEach(s -> s.cleanup(traceId, authorization));
-                    doNetworkEnd(traceId, authorization);
+                    assert cleanupHandler == null;
+                    cleanupHandler = this::doNetworkEnd;
+                    cleanup();
                 }
             }
 
             decoder = decodeIgnoreAll;
         }
 
+        private void onNetworkSignal(
+            SignalFW signal)
+        {
+            if (signal.signalId() == CLEANUP_SIGNAL)
+            {
+                cleanup();
+            }
+            else
+            {
+                assert false;
+            }
+        }
+
+        private void cleanup()
+        {
+            int remaining = config.maxCleanupStreams();
+            for (Iterator<Http2Exchange> iterator = streams.values().iterator();
+                 iterator.hasNext() && remaining > 0; remaining--)
+            {
+                final Http2Exchange stream = iterator.next();
+                stream.cleanup(traceId, authorization);
+            }
+
+            if (!streams.isEmpty())
+            {
+                executor.schedule(100, TimeUnit.MILLISECONDS, routeId, replyId, CLEANUP_SIGNAL);
+            }
+            else
+            {
+                cleanupHandler.accept(traceId, authorization);
+            }
+        }
+
         private void onNetworkAbort(
             AbortFW abort)
         {
-            final long traceId = abort.traceId();
-            final long authorization = abort.authorization();
+            traceId = abort.traceId();
+            authorization = abort.authorization();
```

Comment:
Avoid retaining instance state for `traceId` and `authorization`.

Reply:
We need to retain them since my call `cleanupStreams` onSignal right?